src/Workspaces/Core/Desktop/Workspace/MSBuild/MSBuildWorkspace.cs

```diff
@@ -842,8 +844,25 @@ private class ResolvedReferences
                     {
                         // load the project
                         var projectId = await this.GetOrLoadProjectAsync(fullPath, loader, preferMetadata, loadedProjects, cancellationToken).ConfigureAwait(false);
-                        resolvedReferences.ProjectReferences.Add(new ProjectReference(projectId, projectFileReference.Aliases));
-                        continue;
+
+                        // If that other project already has a reference on us, this will cause a circularity.
+                        // This check doesn't need to be in the "already loaded" path above, since in any circularity this path
+                        // must be taken at least once.
+                        if (ProjectAlreadyReferencesProject(loadedProjects, projectId, targetProject: thisProjectId))
+                        {
+                            // We'll try to make this metadata if we can
+                            var projectMetadata = await this.GetProjectMetadata(fullPath, projectFileReference.Aliases, _properties, cancellationToken).ConfigureAwait(false);
+                            if (projectMetadata != null)
```

Comment:
I'm matching the check from a few lines above. Given this is already a scenario where evil exists and all assumptions should be questioned, it seems best.